aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -1867,15 +1868,30 @@ std::tuple<Tensor, Tensor, Tensor> _lu_with_info_cuda(const Tensor& self, bool p
   Tensor pivots_tensor = at::arange(1, k + 1, self.options().dtype(at::kInt)).expand(req_size).contiguous();
   req_size.pop_back();
   auto infos_tensor = at::zeros(req_size, self.options().dtype(at::kInt));
+  auto batch_size = cuda_int_cast(batchCount(self), "batch size");
 
   Tensor self_working_copy;
   if (self.numel() == 0) {
     self_working_copy = at::empty_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
   } else {
     self_working_copy = cloneBatchedColumnMajor(self);
-    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(self.scalar_type(), "lu_cuda", [&]{
+
+#ifdef USE_CUSOLVER
+    // Heuristic: for single matrix factorization and batch sizes less 8 and matrix sizes less than
+    // 512 we default to looped cuSOLVER if available. Otherwise revert to the default MAGMA function.
+    if (batch_size == 1 || (batch_size <= 8 && m <= 512)) {
```

Comment:
Also add ` || !use_magma_` in the condition.

Otherwise, if someone build pytorch with cusolver but without magma, they can only do LU factorization in the current shapes, but get RTE for "magma not installed" on other shapes, which is weird.